sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -94,6 +106,20 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
     public final PCollection<T> expand(PBegin input) {
       source.validate();
 
+      if (ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api")) {
```

Comment:
Does that mean we can remove https://github.com/apache/beam/blob/e1852ca6af92d61467bfa7dac84e96a2924eefac/runners/portability/java/src/main/java/org/apache/beam/runners/portability/PortableRunner.java#L154 ?

Reply:
We could exclude the override if "beam_fn_api" was used as an experiment.

Reply:
I thought the only application of this override is with the `beam_fn_api` flag enabled. The legacy translation does not support Impulse.

Reply:
Users/framework will need to ensure that the `beam_fn_api` experiment is always used.

Reply:
I was able to remove JavaReadViaImpulse. It required fixing up some tests that weren't setting the `beam_fn_api` experiment.